File: scripts/stager-mon-hunk.py
# Stage UNIQUEMENT les hunks d'un fichier partage qui portent un slug donne.
#
# ⛔ POURQUOI CE SCRIPT (29/08/2026). `git add lib/fiches/registre.ts` prend le
# fichier tel qu'il est et emporte les lignes non commitees des autres sessions.
# La parade documentee — patch applique a l'index — ne suffit pas non plus si
# l'on applique le patch ENTIER : le 29/08, `git diff -U0` rendait deux hunks, et
# le second etait une entree `maths/4e/prop-echelle` de la session maths. Elle est
# partie dans l'index avant d'etre rattrapee.
#
# Usage : python scripts/stager-mon-hunk.py <fichier> <slug>
import subprocess, sys, tempfile, os

fichier, slug = sys.argv[1], sys.argv[2]
diff = subprocess.run(["git", "diff", "-U0", "--", fichier],
                      capture_output=True, text=True, encoding="utf-8").stdout
lignes = diff.splitlines(keepends=True)
entete = lignes[:4]
hunks, cur = [], None
for l in lignes[4:]:
    if l.startswith("@@"):
        if cur:
            hunks.append(cur)
        cur = [l]
    elif cur is not None:
        cur.append(l)
if cur:
    hunks.append(cur)

miens = [h for h in hunks if any(slug in x for x in h)]
autres = len(hunks) - len(miens)
if not miens:
    sys.exit(f"⛔ aucun hunk ne porte « {slug} »")
print(f"{len(hunks)} hunk(s) — {len(miens)} a moi, {autres} laisse(s) intact(s)")

patch = "".join(entete) + "".join("".join(h) for h in miens)
chemin = os.path.join(tempfile.gettempdir(), "mon-hunk.patch")
with open(chemin, "w", encoding="utf-8", newline="") as f:
    f.write(patch)
subprocess.run(["git", "apply", "--cached", "--unidiff-zero", chemin], check=True)
print("applique a l'index")

# ⛔ ET LE CONTROLE QUI MANQUAIT, PAYE LE 29/08 : L'INDEX PEUT SORTIR AVEC UNE
# CLE EN DOUBLE. `git diff -U0` aligne parfois une insertion en REEMETTANT le
# bloc d'ancrage : le hunk dit alors « ajoute dix lignes » la ou l'arbre n'en
# ajoutait que cinq, et `git apply --cached` ecrit les deux. L'arbre de travail
# reste juste, l'index non — et rien ne le signale, parce que TypeScript ne
# refuse pas une cle repetee dans un objet type `Record<string, …>`.
# On relit donc l'index apres coup, et on refuse de rendre la main s'il double.
import re

contenu = subprocess.run(["git", "show", f":{fichier}"],
                         capture_output=True, text=True, encoding="utf-8").stdout
cles = re.findall(r'^  "([^"]+)": \{', contenu, re.M)
doublons = sorted({c for c in cles if cles.count(c) > 1})
if doublons:
    subprocess.run(["git", "restore", "--staged", fichier], check=True)
    sys.exit(f"⛔ l'index sortait avec {len(doublons)} cle(s) en double : {doublons}\n"
             f"   index remis a HEAD. Reprends l'insertion a un autre point d'ancrage.")
print(f"index verifie : {len(cles)} cles, aucune en double")

File: scripts/stager-mon-hunk.mjs
// Stage UNIQUEMENT les hunks d'un fichier partagé qui portent un slug donné.
//
// ⛔ POURQUOI CE SCRIPT (29/08/2026). `git add lib/fiches/registre.ts` prend le
// fichier tel qu'il est et emporte les lignes non commitées des autres sessions.
// La parade documentée — patch appliqué à l'index — ne suffit pas non plus si
// l'on applique le patch ENTIER : `git diff` rendait deux hunks, et le second
// était une entrée `maths/4e/prop-echelle` de la session maths. Elle est partie
// dans l'index avant d'être rattrapée.
//
// ⛔ EN NODE, ET PAS EN PYTHON. La première version était un `.py`. La session
// maths a mesuré que PYTHON N'EST PAS DISPONIBLE dans cet environnement
// (« Python was not found; run without arguments to install from the Microsoft
// Store ») : le script était donc inutilisable depuis une autre session que la
// mienne, ce qui est le pire défaut possible pour un outil de dépôt partagé.
// Node tourne partout ici — c'est déjà lui qui fait tourner les vérificateurs.
//
// Usage : node scripts/stager-mon-hunk.mjs <fichier> <slug>

import { execFileSync } from "node:child_process";
import { writeFileSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

const [fichier, slug] = process.argv.slice(2);
if (!fichier || !slug) {
  console.error("Usage : node scripts/stager-mon-hunk.mjs <fichier> <slug>");
  process.exit(2);
}

const git = (args) => execFileSync("git", args, { encoding: "utf8", maxBuffer: 64 * 1024 * 1024 });

// ⛔ TROIS LIGNES DE CONTEXTE, ET PAS ZÉRO — PAYÉ LE 29/08.
// Avec `-U0`, git n'a aucun point d'ancrage : pour représenter une insertion, il
// lui arrive de RÉÉMETTRE le bloc suivant, si bien qu'un ajout de cinq lignes
// sort en hunk de dix. `git apply --cached` écrit alors les deux, et l'index
// repart avec une clé en double que l'arbre de travail n'a pas. C'est arrivé une
// fois, c'est parti dans un commit poussé, et rien ne l'a signalé.
// Avec `-U3`, le hunk porte son contexte : git apply le POSITIONNE au lieu de le
// deviner, et il ÉCHOUE bruyamment si le fichier a bougé — ce qui est le
// comportement voulu sur un fichier partagé entre trois sessions.
const diff = git(["diff", "-U3", "--", fichier]);
if (!diff.trim()) {
  console.error(`⛔ aucun changement non stagé dans ${fichier}`);
  process.exit(1);
}

const lignes = diff.split(/(?<=\n)/);
const entete = [];
const hunks = [];
let courant = null;
for (const l of lignes) {
  if (l.startsWith("@@")) {
    if (courant) hunks.push(courant);
    courant = [l];
  } else if (courant) {
    courant.push(l);
  } else {
    entete.push(l);
  }
}
if (courant) hunks.push(courant);

const miens = hunks.filter((h) => h.some((x) => x.includes(slug)));
if (miens.length === 0) {
  console.error(`⛔ aucun hunk ne porte « ${slug} »`);
  process.exit(1);
}
console.log(`${hunks.length} hunk(s) — ${miens.length} à moi, ${hunks.length - miens.length} laissé(s) intact(s)`);

const chemin = join(tmpdir(), "mon-hunk.patch");
writeFileSync(chemin, entete.join("") + miens.map((h) => h.join("")).join(""), "utf8");
git(["apply", "--cached", chemin]);
console.log("appliqué à l'index");

// ⛔ ET LE CONTRÔLE QUI MANQUAIT, PAYÉ LE 29/08 : L'INDEX PEUT SORTIR AVEC UNE
// CLÉ EN DOUBLE, et rien ne le signale — TypeScript ne refuse pas une clé
// répétée dans un objet typé `Record<string, …>`, donc `tsc --noEmit` passe.
// On relit donc l'index après coup, et on refuse de rendre la main s'il double.
const contenu = git(["show", `:${fichier}`]);
const cles = [...contenu.matchAll(/^ {2}"([^"]+)": \{/gm)].map((m) => m[1]);
const vus = new Map();
for (const c of cles) vus.set(c, (vus.get(c) ?? 0) + 1);
const doublons = [...vus].filter(([, n]) => n > 1).map(([c]) => c).sort();
if (doublons.length) {
  git(["restore", "--staged", fichier]);
  console.error(
    `⛔ l'index sortait avec ${doublons.length} clé(s) en double : ${doublons.join(", ")}\n` +
      `   index remis à HEAD. Reprends l'insertion à un autre point d'ancrage.`
  );
  process.exit(1);
}
console.log(`index vérifié : ${cles.length} clés, aucune en double`);
